src/modules/claim/claim.service.ts

```diff
@@ -75,73 +69,149 @@ export class ClaimService {
+
+    const dto = await ClaimRequestDTO.create({
+      ...rq,
+      claimType,
+      claimTypeVersion,
+    });
+
+    await this.roleService.verifyEnrolmentPrecondition({
+      claimType,
+      userDID: dto.requester,
+    });
+
+    await this.create(dto);
+  }
+
+  /**
+   * Handles claim issuance request saving and updates.
+   * Two scenarios are handled - issue requested claim and issue not-requested claim
+   * @param rq IClaimIssuance request
    */
-  public async handleExchangeMessage(
-    data: IClaimIssuance | IClaimRejection | IClaimRequest,
-  ): Promise<string> {
-    try {
-      const claim: RoleClaim = await this.getById(data.id);
-      if (!claim && 'token' in data) {
-        const {
-          claimData: { claimType, claimTypeVersion },
-        } = jwt.decode(data.token) as DecodedClaimToken;
-
-        const dto = await ClaimRequestDTO.create({
-          ...data,
-          claimType,
-          claimTypeVersion,
-        });
-
-        await this.roleService.verifyEnrolmentPrecondition({
-          claimType,
-          userDID: dto.requester,
-        });
-
-        await this.create(dto);
-        return;
-      }
-      if (!claim && 'issuedToken' in data) {
-        const {
-          claimData: { claimType, claimTypeVersion },
-        } = jwt.decode(data.issuedToken) as DecodedClaimToken;
-
-        const dto = await NewClaimIssueDTO.create({
-          ...data,
-          claimType,
-          claimTypeVersion,
-        });
-
-        await this.roleService.verifyEnrolmentPrecondition({
-          claimType,
-          userDID: dto.requester,
-        });
-
-        await this.createAndIssue(dto);
-        return;
-      }
-      if (claim && !claim.isAccepted && 'isRejected' in data) {
-        const dto = await ClaimRejectionDTO.create(data);
-
-        await this.reject(dto.id);
-        return;
-      }
-
-      if (
-        (claim && !claim.isRejected && 'issuedToken' in data) ||
-        'onChainProof' in data
-      ) {
-        const dto = await ClaimIssueDTO.create(data);
-
-        await this.issue(dto);
-        return;
-      }
-    } catch (err) {
-      this.logger.error(err);
+  public async handleClaimIssuanceRequest(rq: IClaimIssuance) {
+    const claim: RoleClaim = await this.getById(rq.id);
+    const {
+      claimData: { claimType, claimTypeVersion },
+    } = jwt.decode(rq.issuedToken) as DecodedClaimToken;
+
+    if (!claim && !!rq.issuedToken) {
+      const dto = await NewClaimIssueDTO.create({
+        ...rq,
+        claimType,
+        claimTypeVersion,
+      });
+
+      await this.roleService.verifyEnrolmentPrecondition({
+        claimType,
+        userDID: dto.requester,
+      });
+
+      await this.createAndIssue(dto);
+    } else if (
+      claim &&
+      !claim.isRejected &&
+      (!!rq.issuedToken || !!rq.onChainProof)
+    ) {
+      const dto = await ClaimIssueDTO.create(rq);
+      await this.issue(dto);
     }
```

Comment:
What do you think of returning an error here so that an error HTTP code can be returned to the client? I think it would be weird to continue to publish the message at this point is the issuance storage hasn't taken place